src/parse/parser.rs

```diff
@@ -1467,89 +1467,98 @@ where
-                            $_self.add_val_to_arg($a, val, $m)?;
-                        }
+                    if add {
+                        self.add_val_to_arg(arg, OsStr::new(default), matcher)?;
+                        done = true;
+                        break;
                     }
-                } else {
-                    debugln!(
-                        "Parser::add_defaults:iter:{}: doesn't have default vals",
-                        $a.name
-                    );
                 }
-            };
-            ($_self:ident, $a:ident, $m:ident) => {
-                if let Some(ref vm) = $a.default_vals_ifs {
-                    sdebugln!(" has conditional defaults");
-                    let mut done = false;
-                    if $m.get(&$a.id).is_none() {
-                        for (arg, val, default) in vm.values() {
-                            let add = if let Some(a) = $m.get(&arg) {
-                                if let Some(v) = val {
-                                    a.vals.iter().any(|value| v == value)
-                                } else {
-                                    true
-                                }
-                            } else {
-                                false
-                            };
-                            if add {
-                                $_self.add_val_to_arg($a, OsStr::new(default), $m)?;
-                                done = true;
-                                break;
-                            }
-                        }
-                    }
+            }
 
-                    if done {
-                        continue; // outer loop (outside macro)
-                    }
-                } else {
-                    sdebugln!(" doesn't have conditional defaults");
-                }
-                add_val!(@default $_self, $a, $m)
-            };
+            if done {
+                return Ok(());
+            }
+        } else {
+            sdebugln!(" doesn't have conditional defaults");
         }
 
-        for o in opts!(self.app) {
-            debug!("Parser::add_defaults:iter:{}:", o.name);
-            add_val!(self, o, matcher);
-        }
-        for p in positionals!(self.app) {
-            debug!("Parser::add_defaults:iter:{}:", p.name);
-            add_val!(self, p, matcher);
+        if let Some(ref vals) = arg.default_vals {
+            debugln!("Parser::add_defaults:iter:{}: has default vals", arg.name);
+            if matcher
+                .get(&arg.id)
+                .map(|ma| ma.vals.len())
+                .map(|len| len == 0)
+                .unwrap_or(false)
+            {
+                debugln!(
+                    "Parser::add_defaults:iter:{}: has no user defined vals",
+                    arg.name
+                );
+
+                for val in vals {
+                    self.add_val_to_arg(arg, val, matcher)?;
+                }
+
+                Ok(())
```

Comment:
Please extract the `Ok(())` in this conditionals out since it is common for all the conditionals.